tensorflow/python/keras/engine/data_adapter.py

```diff
@@ -1412,17 +1412,52 @@ def _make_class_weight_map_fn(class_weight):
     A function that can be used with `tf.data.Dataset.map` to apply class
     weighting.
   """
-  class_ids = list(sorted(class_weight.keys()))
+  # Convert string keys to integers
+  converted_class_weight = {}
+  for key, value in class_weight.items():  # pylint: disable=unused-variable
+    # Check for complex values first
+    try:
+      tensor_val = tensor_conversion.convert_to_tensor_v2_with_dispatch(value)
+      if tensor_val.dtype.is_complex:
+        raise ValueError(
+            f"Complex class weights are not supported. "
+            f"Found complex value {value} for class {key}. "
+            f"Please use real-valued weights only.")
+    except Exception:
+      # If we can't convert to tensor, let the original error handling deal with it
+      pass
+    
+    # Convert keys to integers
+    if isinstance(key, (int, numpy_compat.integer_types)):
+      converted_class_weight[key] = value
+    elif isinstance(key, str):
+      try:
+        int_key = int(key)
+        converted_class_weight[int_key] = value
+      except ValueError:
+        raise ValueError(f"Invalid class_weight key: '{key}'. "
+                        f"Class weight keys must be integers representing "
+                        f"class indices, "
+                        f"got key of type {type(key).__name__}.")
+    else:
+      raise ValueError(f"Invalid class_weight key: '{key}'. "
+                      f"Class weight keys must be integers representing "
+                      f"class indices, "
+                      f"got key of type {type(key).__name__}.")
+
+  # Validate class indices are contiguous starting from 0
+  class_ids = sorted(converted_class_weight.keys())
   expected_class_ids = list(range(len(class_ids)))
   if class_ids != expected_class_ids:
     error_msg = (
         "Expected `class_weight` to be a dict with keys from 0 to one less "
         "than the number of classes, found {}").format(class_weight)
     raise ValueError(error_msg)
-
+   
```

Comment:
You still have trailing whitespace. Please configure your editor to strip in on save, I have to mention this in almost every round of review.